python_modules/libraries/dagster-dbt/dagster_dbt/asset_decorators.py

```diff
@@ -0,0 +1,86 @@
+from dataclasses import dataclass
+from functools import cached_property
+from typing import Callable, Mapping, Optional
+
+from dagster import AssetsDefinition, multi_asset
+from dagster._core.definitions.events import AssetKey
+
+from .asset_utils import default_asset_key_fn, get_dbt_multi_asset_args, get_deps
+from .utils import ASSET_RESOURCE_TYPES, output_name_fn, select_unique_ids_from_manifest
+
+
+@dataclass
+class DbtManifest:
```

Comment:
Is this class foreshadowing future utility functions that would take a DbtManifest as input? If so, what are patterns you'd imagine using it in?

It seems like it's not necessary for this PR (just because only the `dbt_unique_id_to_nodes` property is used, so it might be better to punt on this thing until it's needed.

Reply:
I'm imagining that the user will use this class for customization:

- When customizing the asset keys of the `AssetsDefinition` after invoking `@dbt_assets`. Here, we provide them a way to access the node info so they can map it however they please (e.g. to generate a custom asset key).
- Also, this allows the user to have access to the manifest if they want to customize the body of their computation (e.g. adding arbitrary metadata to their `AssetMaterialization`/`AssetObservation`/`Output`), using either the asset key or output.
- I am also envisioning a similar class for dbt's Run Results artifact.

Reply:
My feeling is that if the main usecase of this class is to help users translate between dbt representations and dagster representations of an asset, we'd want to bundle it together with the asset key generating function. there are a few independent places where we'd like to be able to do this translation:

- doing the `with_attributes` call after `@dbt_assets`
- inside the compute_fn, to potentially do fancy things
- when generating AssetObservations for tests (we need to covert the upstream unique ids for the test into upstream asset keys for the observations)
- when generating an asset selection via `DbtManifestAssetSelection`

There's also a particularly rough edge when you involve key prefixes, as these can get applied post-hoc, meaning there's no automatic way of ensuring these callsites are in sync with whatever key prefix has been added. Not sure if there's any perfect solution for that bit, but it's something to be aware of

Anyway, I think it'd be an ok outcome if the user was able to create their own subclass of this object to supply their own definition for `get_asset_key_for_unqiue_id` or something, I just feel like we probably wouldn't want to call this object `DbtManifest` in that case, as it's more of a translation layer